### PR TITLE
refactor: parsing and validation

### LINKED ISSUE #3371: Add tests to cover the negative API requests

### PR DESCRIPTION
**Pull request by swarmHaseeb**
### Checklist

- [x] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [ ] My change requires a documentation update, and I have done it.
- [ ] I have added tests to cover my changes.
- [x] I have filled out the description and linked the related issues.

### Description

parsing and validation of headers and changes in tests 
w thanks to Vlado
#### Motivation and Context (Optional)
extension of #3344 

### Related Issue (Optional)

related #3163 #3165 #3166 
blocking #3249 #3344 
closes #3371

### CODE REVIEW

**Comment by swarmHaseeb** [comment]
Require input for current approach. If this looks good, i can continue on it and adjust code as necessary and needed.
@mrekucci apart from validation approach can you please check if the TODOs were fixed correctly?
@vladopajic @umerm-work for suggestions/reviews

File: pkg/api/api.go [hunk 1]
```diff
@@ -67,6 +67,148 @@ import (
 // loggerName is the tree path name of the logger for this package.
 const loggerName = "api"
 
+type headerDescriptor struct {
+	// name of header parameter
+	Name string
+
+	// Validate will validate parameter value or
+	// return error if value is not in correct format.
+	Validate func(v interface{}, s *Service) error
```
Comment (mrekucci) [inline pkg/api/api.go:76]: Why is the `v` interface{} type when headers are always strings?

File: pkg/api/api.go [hunk 2]
```diff
@@ -67,6 +67,148 @@ import (
...
+var (
+	swarmPinHeader = headerDescriptor{
+		Name: "Swarm-Pin",
+		Validate: func(v interface{}, _ *Service) error {
+			v = strings.ToLower(v.(string))
+			switch v {
+			case v.(*bool) == nil || !v.(bool) || v.(bool):
+				return nil
+			default:
+				return fmt.Errorf("%w: expected bool", errInvalidHeaderParams)
```
Comment (mrekucci) [inline pkg/api/api.go:91]: What is the purpose of wrapping the `errInvalidHeaderParams`, where it is unwrapped? The same is below.

File: pkg/api/api.go [hunk 3]
```diff
@@ -567,25 +703,32 @@ func (s *Service) contentLengthMetricMiddleware() func(h http.Handler) http.Hand
...
-			}{}
-			if response := s.mapStructure(r.Header, &headers); response != nil {
-				response("invalid header params", logger, w)
-				return
-			}
 			ctx := r.Context()
-			ctx = sctx.SetGasPrice(ctx, headers.GasPrice)
-			ctx = sctx.SetGasLimit(ctx, headers.GasLimit)
 
+			for _, hd := range allHeaderDescriptors {
```
Comment (mrekucci) [inline pkg/api/api.go:714]: The iteration, validation, and context setting of all headers must be reflected in each endpoint described in the Open API specification, including a description of what the headers are used for. Since most endpoints do not use all of these headers, this is not the correct approach. The Open API specification should be on pair with the implementation.

File: pkg/api/util.go [hunk 4]
```diff
@@ -146,7 +149,7 @@ var preMapHooks = map[string]func(v string) (string, error){
 //
 // In case of parsing error, a new parseError is returned to the caller.
 // The caller can use the Unwrap method to get the original error.
-func mapStructure(input, output interface{}) (err error) {
+func mapStructure(input, output interface{}, s *Service) (err error) {
```
Comment (mrekucci) [inline pkg/api/util.go:152]: This is a generic function, it should not take a `Service` object as a param, but the object should be wrapped in a pre-hook closure.

**Review by mrekucci — COMMENTED** [review]
I find the validators to be very limiting, not only is it hard to configure or indicate that some headers are required and some are not, but the output errors are very generic, they don't tell why the validation failed, e.g.: odd hexadecimal characters, not a hexadecimal string, etc...

File: pkg/api/util.go (hunk 4 above)
Comment (swarmHaseeb) [reply pkg/api/util.go:152]: thankyou, yeah makes sense, on it already

File: pkg/api/api.go (hunk 3 above)
Comment (swarmHaseeb) [reply pkg/api/api.go:714]: i had a discussion about this same issue with @vladopajic on Friday or Sat and thanks to him there is an approach we can use to validate that and it also helps with actually enforcing some rules for open api docs and based on this approach i also have found a solution for it will also be sharing that later today probably, maybe we can explore it slightly more and see if it sorts out all those problems we had prev cuz like an extension of this solution will also fix the research part that we were figuring out before.
I am improving the error messages to be more user/dev friendly.

File: pkg/api/api.go (hunk 2 above)
Comment (swarmHaseeb) [reply pkg/api/api.go:91]: sorry it's still a draft pr so i'm playing around with different things. Hopefully will have some proper implementation by tomorrow.

File: pkg/api/api.go (hunk 1 above)
Comment (swarmHaseeb) [reply pkg/api/api.go:76]: sorry, i was playing around and focused on other part and didn't give it much thought before. Will be fixed in next commit.

File: pkg/api/api.go [hunk 5]
```diff
@@ -67,6 +67,145 @@ import (
...
+}
+
+var (
+	swarmPinHeader = headerDescriptor{
+		Name: "Swarm-Pin",
+		Validate: func(v string, _ *Service) error {
+			v = strings.ToLower(v)
+			if v == boolHeaderSetValue || v != boolHeaderSetValue || v == "" {
+				return nil
+			}
```
Comment (vladopajic) [inline pkg/api/api.go:89]: `v != boolHeaderSetValue` is not correct

--- 
condition should be:

v == "" || v == "true" || v == "false" 

---


this validate function seam se to repeating, could we reuse it in other places?

Comment (swarmHaseeb) [reply pkg/api/api.go:89]: linter was not letting this repeat func go through but maybe i can write another func for it

Comment (vladopajic) [reply pkg/api/api.go:89]:
```
func validateBool(v string, _*Service) error {
 // todo
 return nil
}


...
 headerDescriptor{
	Name: "Swarm-Pin",
	Validate: validateBool,
}
...

```

would this work?

File: pkg/api/bzz.go [hunk 6]
```diff
@@ -128,10 +128,14 @@ func (s *Service) fileUploadHandler(logger log.Logger, w http.ResponseWriter, r
...
-		// TODO: return standard validation error.
+		valErr := &validationError{
+			Entry: "queries.FileName",
+			Value: queries.FileName,
+			Cause: err,
+		}
 		logger.Debug("bzz upload file: / in prefix not allowed", "file_name", queries.FileName, "error", err)
 		logger.Error(nil, "bzz upload file: / in prefix not allowed", "file_name", queries.FileName)
-		jsonhttp.BadRequest(w, "/ in prefix not allowed")
+		jsonhttp.BadRequest(w, valErr)
```
Comment (notanatol) [inline pkg/api/bzz.go:138]: maybe for `bad request` we should not just pass on the technical error but use a plain string instead

Comment (swarmHaseeb) [reply pkg/api/bzz.go:138]: If there is a validation error, if would write a resp like this
```
got json response "{\"code\":400,\"message\":\"invalid query params\",\"reasons\":[{\"field\":\"name\",\"error\":\"want startsnotwith:/\"}]}"
```

**Review by notanatol — APPROVED** [review]

**Review by vladopajic — APPROVED** [review]
